Add WM status LED

hallgchris opened the pull request:
## Description

This PR adds code to control the WM's status LED. The status LED basically indicates what state the WM is in, e.g. connecting to MQTT broker or publishing data.

The status LED can be any of the 8 colour combinations made by having each of red, green and blue on or off. We could use PWM to allow many more colours but we probably don't want to have colours that are too similar, so this is fine for now (i.e. less work for me). The LED can also be blinking or solid. Take care that the LED cannot blink during blocking operations, I've made sure that this doesn't occur in the current implementation.

You can see the pre-defined LED states in `status_led.py`. I'll make a guide on notion for using the back WM shortly, which will include this information in a more reader-friendly manner.

Also changed in this PR:
- Battery and data publishing loops are now in their own `uasyncio` tasks, so that they run fully independently and don't block other tasks (e.g. LED blinking) from running when they're idle.
- The WM will now repeatedly try to connect to MQTT until it is unsuccessful. Previously it would just give up after one unsuccessful attempt.

## Screenshots

No screenshots, but here's the LED
![image](https://user-images.githubusercontent.com/17876556/133868484-f35731a6-e827-48f3-945e-1b6215a82d61.png)

## Steps to Test

If you happen to have the WM somehow, load the code with
```bash
cd wireless_modules/back_module
./upload.sh /dev/ttyUSB0 # substitute correct device
```

then reboot the WM. You should observe statuses as in `status_led.py`.

hallgchris committed:
Add solid status LED states

hallgchris committed:
Attempt reconnection to MQTT broker until successful

hallgchris committed:
Add blinking status LED support

hallgchris committed:
Add low battery warning LED

hallgchris committed:
Add red LED for error state

hallgchris committed:
Tidy up some code I forgot to revert

hallgchris committed:
Change sensor init colour

hallgchris committed:
Fix crash when setting led off

hallgchris committed:
Update example config pins

kdav108 commented on wireless_modules/boot.py:
```diff
@@ -1,9 +1,10 @@
 from machine import Pin
 
+import config
+from status_led import StatusLed, WmState
```
Nit: Would be nice to group the import statements, seeing as we do that in this repo

hallgchris replied:
Aren't thney grouped? They're split into micropython modules and our own files

kdav108 replied:
I believe we're grouping them on the style, all the `import` statements are together, and the `from ... import ...` are grouped together

hallgchris replied:
I think that's just a coincidence that that's also been the case. My intention has always been to group first python built in modules, then 3rd party modules, then our own. This is what we've done in other repos (where we remember to) also.

kdav108 commented on wireless_modules/mqtt_client.py:
```diff
@@ -22,8 +24,24 @@ def connect_and_subscribe(self, topics_to_subscribe, callback_func):
 
         # Connect to MQTT broker
-        self.client.connect()
+        connection_success = False
+        while not connection_success:
+            connection_success = True
```
Bit pedantic, but since we already have the class variable `self.connected`, we can probably just use that here.

kdav108 commented on wireless_modules/status_led.py:
```diff
@@ -0,0 +1,82 @@
+                    self.__set_leds_off()
+                    await asyncio.sleep_ms(200)
+
+            else:
```
What about making this more explicit:`elif self.state` so if we ever call this function without setting any state, the WM won't crash. A very rare scenario, but good to make it crash proof I guess.

hallgchris replied:
Actually, that's why I had the `Undefined` LedState... Just forgot to use it!

hallgchris replied:
Nevermind, I did use it... This state will never happen, unless we explicitly try to set the state to `None`, in which case we want to know about it!

kdav108 replied:
It can happen if we initialise `StatusLed` and do not call `set_state` before calling `start_blink_loop`. Although, that case is an error as well, so I'm fine if we leave it as is

hallgchris replied:
`set_state` is called by `StatusLed.__init__` :)

kdav108 replied:
Ahhh don't know how I missed that

kdav108 commented on wireless_modules/wireless_module.py:
```diff
@@ -73,6 +75,26 @@ def sub_cb(self, topic, msg):
+                    self.battery_topic, ujson.dumps(battery_voltage), retain=True
+                )
+            self.status_led.set_warning_state(
+                WmState.LowBattery if battery_voltage["voltage"] <= 3.3 else None
```
Can we make the low battery threshold (`3.3`) into a variable

kdav108 reviewed (changes requested):
Great design! Using `ayncio` is quite neat!

Unfortunately couldn't test it physically, but took some extra time to simulate it mentally, and it works well :)

kdav108 commented:
Also, if you get the notion manual written up for the back WM explaining the status LED messages before merging this PR, it might be good to reference it from the README.

hallgchris commented:
Note to self before I merge, the magenta for the sensor initialising state is too close to blue. Need to try another colour. Also, the LED doesn't blink despite being told to because the initialisation code is blocking, should make it explicitly not blinking.

hallgchris committed:
Use existing self.connected variable

hallgchris committed:
Extract low battery magic number to constant

hallgchris committed:
Add link to WM user guide to readme

hallgchris committed:
Tidy up WM readme

kdav108 reviewed (approved):
LGTM!

hallgchris committed:
Fix some MQTT connection issues
- Fix not catching MQTTException
- Add missing variable
- Add compatibility with mosquitto 2.0.12

hallgchris committed:
Fix: Reset on errors that are caught by uasyncio

hallgchris committed:
Change sensor initialisation led colour to cyan

hallgchris committed:
Format code with black

hallgchris commented:
Made some changes. I accidentally caused an issue with MQTT connection in the last review round, which made me realise some types of errors aren't handled properly. Also discovered that the latest mosquitto version doesn't play nicely with the `umqtt` modules default settings: see https://github.com/micropython/micropython-lib/issues/445

Finally, also ran into a bug which I believe is introduced in this PR (come to think of it, haven't checked `master`). Leaving the sensor idling for a few minutes causes a funky exception to occur. I can't quite track down the cause, I wonder if it's to do with `umqtt` not being ideal for async code in the first place (source: https://forum.micropython.org/viewtopic.php?t=9581). Might see if using `mqtt_as` as suggested is any different. Error is below, if you have ideas make them heard:

```
Task exception wasn't retrieved
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "wireless_module.py", line 149, in start_data_loop
  File "wireless_module.py", line 127, in wait_for_start
  File "mqtt_client.py", line 77, in check_for_message
  File "umqtt/simple.py", line 204, in check_msg
  File "umqtt/simple.py", line 173, in wait_msg
OSError: -1
```

I haven't been able to find out what `OSError: -1` means.

Otherwise if all else is fine I'd like to get this in and will solve this bug in another PR since it might require a substantial bit of rewriting if we switch MQTT module.

hallgchris committed:
Send ping to MQTT broker once a minute

hallgchris commented:
So turns out the problem was the broker was disconnecting us because `umqtt` doesn't respond to the server's pings, nor does it automatically send its own. This wasn't an issue previously because we were configured not to send pings at all, but with mosquitto 2.0.12 this is no longer an allowable default.

The solution was relatively simple in the end. `umqtt` has a method to manually ping the server, so calling that once a minute works as expected.

kdav108 commented on wireless_modules/wireless_module.py:
```diff
@@ -38,6 +42,22 @@ def __init__(self, module_id, battery_reader=None):
+    def error_handler(self, _loop, context):
+        self.status_led.set_state(WmState.Error)
+        print("An error occured in a uasyncio task!")
+        print(context["message"])
+        sys.print_exception(context["exception"])
```
Do we need both these print statements?

hallgchris replied:
Yep, so by memory the first one will print whatever message is associated with the exception (e.g. `Task exception wasn't retrieved`, looking at the example above) and then the next line prints the actual traceback.

kdav108 replied:
I thought tracebacks themselves print out the actual message as well

kdav108 commented on wireless_modules/wireless_module.py:
```diff
@@ -38,6 +42,22 @@ def __init__(self, module_id, battery_reader=None):
+        print(context["message"])
+        sys.print_exception(context["exception"])
+
+        # FIXME: A nicer solution would be to restart all running tasks rather than reset
```
Is there an easy way to gracefully stop all tasks before reseting the esp32?

hallgchris replied:
Ordinarily there would be, but asyncio doesn't let you "cancel" a task from within the task itself. This function is considered by asyncio to be part of the task that the error occurred in, so unfortunately there's not a simple clean way to cancel the tasks.

To be honest, there's no harm in resetting the esp32 without notice anyway, unlike for example a raspberry pi which mightn't be as big of a fan. When we turn off the esp32 we don't gracefully stop tasks either, we just cut power.

The main reason it would be good to restart tasks rather than reset would be that we could come back online faster and maintain state such as whether we should be logging. However, if it's something we decide we want I think it could go in its own PR, kinda unrelated to the status LED at this point. Another solution to the problem would be to just make the `start`/`stop` topics retained (I reckon we should)

kdav108 replied:
Yep agree `start/stop` should be retained and completely fine with the reset task to be a seperate PR. I didn't see any benefit in gracefully stopping either, it just seemed nice 😂. Hence my initial emphasise on whether `there is an easy way` to do it.

kdav108 commented on wireless_modules/wireless_module.py:
```diff
@@ -38,6 +42,22 @@ def __init__(self, module_id, battery_reader=None):
+
+        # FIXME: A nicer solution would be to restart all running tasks rather than reset
+
+        print("Resetting in 5 seconds...")
+
+        time.sleep(5)
+        machine.reset()
```
We already catch exceptions and do this in `back_module/main.py`, if errors thrown by async functions reach that `try and catch` block, we can just rely on that to reset the machine.

hallgchris replied:
Good catch, but actually asyncio catches and suppresses the errors, even though it may kill the tasks. It does print the traceback by default, but won't trigger a reset meaning the WM is out of action until manually power cycled.

kdav108 replied:
So does that mean the try catch block is redundant in the `main.py` script?

hallgchris replied:
The try catch there is still useful in case the initialisation crashes, that won't be caught by asyncio.

kdav108 commented on wireless_modules/wireless_module.py:
```diff
@@ -2,22 +2,26 @@
 except FileNotFoundError:
     print("Error importing config.py, ensure a local version of config.py exists")
 
+LOW_BATTERY_THRESHOLD = 3.3
```
It's probably not a concern for the short-medium term, but is it possible in the future for different WM to use different types of batteries and hence have different low battery thresholds?

hallgchris replied:
Very very unlikely. All lithium chemistry batteries have a 3.7 volt nominal voltage and it'd be fairly safe to say that 3.3 will always be a good threshold (or if we decide it's too low or too high, we'd want to adjust it for all WMs). Only possibility of a different voltage is if we use multiple cells, but currently we're looking at significantly reduced power consumption in the front WM, not increased. I think this is fine.

kdav108 reviewed (changes requested):
Thanks for fixing these issues (and picking them up int he first place 😂)! 

Just a few suggestions, may not even need to entertain all of them...

hallgchris commented:
No changes made lol, easiest review response ever. Let me know your thoughts

kdav108 reviewed (approved):
Happy days!
